src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py

```diff
@@ -267,31 +315,106 @@ def _get_partitioned_plan(self, root: TreeNode, child: TreeNode) -> SnowflakePla
+            is_valid = False
+            validity_status = InvalidNodesInBreakdownCategory.SCORE_BELOW_LOWER_BOUND
+
+        if score > self.complexity_score_upper_bound:
+            is_valid = False
+            validity_status = InvalidNodesInBreakdownCategory.SCORE_ABOVE_UPPER_BOUND
+
+        if is_valid and not self._is_node_pipeline_breaker(node):
+            is_valid = False
+            validity_status = InvalidNodesInBreakdownCategory.NON_PIPELINE_BREAKER
+
+        if is_valid and self._contains_external_cte_ref(node, root):
+            is_valid = False
+            validity_status = InvalidNodesInBreakdownCategory.EXTERNAL_CTE_REF
+
+        if is_valid:
             _logger.debug(
                 f"Added node of type {type(node)} with score {score} to pipeline breaker list."
             )
-        return valid_node, score
+
+        return validity_status, score
+
+    def _contains_external_cte_ref(self, node: TreeNode, root: TreeNode) -> bool:
+        """Method to check if a node contains a CTE in its subtree that is also referenced
+        by a different node that lies outside the subtree. An example situation is:
+
+                                   root
+                                /       \
+                            node1       node5
+                            /    \
+                        node2    node3
+                       /    |      |
+                   node4  SelectSnowflakePlan
+                                |
+                           SnowflakePlan
+                                |
+                           WithQueryBlock
+                                |
+                              node6
+
+        In this example, node2 contains a WithQueryBlock node that is also referenced
+        externally by node3.
+        Similarly, node3 contains a WithQueryBlock node that is also referenced externally
+        by node2.
+        However, node1 contains WithQueryBlock node that is not referenced externally.
+
+        If we compare the count of WithQueryBlock for different nodes, we get:
+          NODE:                 COUNT:    Externally Referenced:
+          ======================================================
+          node1                 2         False
+          node2                 1         True
+          node3                 1         True
+          root                  2         False
+          SelectSnowflakePlan   1         False
+          SnowflakePlan         1         False
+
+        We determine if a node contains an externally referenced CTE by comparing the
+        number of times each unique WithQueryBlock node is referenced in the subtree compared
+        to the number of times it is referenced in the root node.
+        """
+
+        # Checks for SnowflakePlan and SelectSnowflakePlan is to prevent marking a WithQueryBlock, which is a pipeline breaker
+        # node as an external CTE ref.
+        if isinstance(node, SelectSnowflakePlan):
```

Comment:
here we need a special handling is because we want to apply the check on the snowflakePlan to handle the cte case mentioned below